VSR/SuperBlock: Trailer IO dedup

sentientwaffle opened the pull request:
Also:
- Move the request/response command mapping from `superblock.Trailer` to `sync.Trailer`. (More relevant).
- Split `SuperBlock.trailer()` into `SuperBlock.trailer_buffer`, `SuperBlockHeader.trailer_checksum`, and `SuperBlockHeader.trailer_size`. (`.trailer()` previously always used `.staging` as its header, but now that isn't always what we want. This is more explicit too.)

The diff in `superblock.zig` makes this refactor look much more extensive than it is.

sentientwaffle committed:
VSR/SuperBlock: Dedup trailer write path

Move the request/response command mapping from `superblock.Trailer` to `sync.Trailer`. (More relevant).

sentientwaffle committed:
VSR/SuperBlock: Dedup trailer read path

matklad commented on src/vsr/superblock.zig:
```diff
@@ -1823,37 +1741,33 @@ pub const Area = enum {
-        .base = free_set.base + free_set.size_max,
-        .size_max = superblock_trailer_client_sessions_size_max,
-    };
+    pub fn base(area: Area) u64 {
```
Let's use the same `inline for` here we use for `vsr.Zone`. Also, maybe call this `SuperblockZone`? I think `Zone` and `Area` are the same kind of thing, so using different words for two feels redundant. 

Ah, we also have testing/storage/Area, and that is vsr.Zone + a specific entry in the zone. So, we have Zone and two Areas, and the semantics between the two Areas is different :)

Yeah, it feels like this should be `SuperBlockZone` or maybe `SuperBlock.Zone`

matklad replied:
I am also not happy with base/offset naming here. `superblock.Area.offset` is `vsr.Zone.start`, and `vsr.Zone.offset` is something completely different. And, well, in 

```
return superblock_copy_size * @as(u64, copy) + area.base();
```

`superblock_copy_size * @as(u64, copy)` is base, and `area.base()` is offset :rofl: 

maybe `SuperBlockZone.start` and `SuperBlockZone.start_for_copy()`?

matklad commented on src/vsr/superblock.zig:
```diff
@@ -978,133 +979,60 @@ pub fn SuperBlockType(comptime Storage: type) type {
+            switch (trailer_) {
+                .manifest => context.superblock.write_trailer(context, .free_set),
+                .free_set => context.superblock.write_trailer(context, .client_sessions),
+                .client_sessions => context.superblock.write_header(context),
```
Awesome simplification here!

matklad commented on src/vsr/superblock.zig:
```diff
@@ -1823,37 +1741,33 @@ pub const Area = enum {
-        base: u64,
-        size_max: u64,
+    pub fn offset(area: Area, copy: u8) u64 {
+        return superblock_copy_size * @as(u64, copy) + area.base();
```
assert that `copy < superblock_copies`

matklad commented on src/vsr/superblock.zig:
```diff
@@ -978,133 +979,60 @@ pub fn SuperBlockType(comptime Storage: type) type {
-            assert(superblock.staging.manifest_checksum == vsr.checksum(
-                superblock.manifest_buffer[0..superblock.staging.manifest_size],
-            ));
+            const offset = trailer.area().offset(context.copy.?);
```
Move offset computation below zerroing-out the buffer, right now is looks like `offset` is related to buffer, but it is not.

matklad commented on src/vsr/superblock.zig:
```diff
@@ -924,7 +925,7 @@ pub fn SuperBlockType(comptime Storage: type) type {
             context.copy = 0;
             if (context.caller.updates_trailers()) {
-                superblock.write_manifest(context);
+                superblock.write_trailer(context, .manifest);
```
for clarity, maybe add `superblock.write_trailers()` which delegates to `write_trailer` (which we could rename to `write_next_trailer` to ehphasize that it writes a chain of trailers, rather than just one?)

matklad commented on src/vsr/superblock.zig:
```diff
@@ -978,133 +979,60 @@ pub fn SuperBlockType(comptime Storage: type) type {
 
-            const buffer = superblock.manifest_buffer[0..size];
-            const offset = areas.manifest.offset(context.copy.?);
+            const trailer_data_ = superblock.trailer_data(trailer);
```
Can we make any area-specific assertions here? Eg, `if area == client_sessions { assert(data.size % ??? == 0)}`?

matklad commented on src/testing/storage.zig:
```diff
@@ -502,7 +502,7 @@ pub const Storage = struct {
     ) *const superblock.SuperBlockHeader {
         const offset = vsr.Zone.superblock.offset(superblock.Area.header.offset(copy_));
-        const bytes = storage.memory[offset..][0..superblock.Area.header.size_max()];
+        const bytes = storage.memory[offset..][0..comptime superblock.Area.header.size_max()];
```
:zap:

matklad commented on src/vsr/replica.zig:
```diff
@@ -8192,16 +8192,18 @@ pub fn ReplicaType(
+            const trailer_buffer_all = self.superblock.trailer_buffer(trailer);
+            const trailer_checksum = self.superblock.staging.trailer_checksum(trailer);
+            const trailer_size = self.superblock.staging.trailer_size(trailer);
+            assert(trailer_size >= parameters.offset);
```
Should this be a strict inequality? Doesn't make sense to request empty chunk

matklad commented on src/vsr/replica.zig:
```diff
@@ -8192,16 +8192,18 @@ pub fn ReplicaType(
+                trailer_size - parameters.offset,
                 constants.sync_trailer_message_body_size_max,
             ));
             assert(body_size <= constants.message_body_size_max);
```
similarly, assert that body_size > 0

sentientwaffle replied:
`body_size` can be zero in the case of e.g. an empty manifest -- when offset is 0 the syncing replica isn't just requesting the body, they are also discovering the size and checksum of the whole trailer.

I'll add `assert(body_size > 0 or parameters.offset == 0);`.

matklad commented on src/vsr/replica.zig:
```diff
@@ -8192,16 +8192,18 @@ pub fn ReplicaType(
+            const trailer_buffer_all = self.superblock.trailer_buffer(trailer);
+            const trailer_checksum = self.superblock.staging.trailer_checksum(trailer);
+            const trailer_size = self.superblock.staging.trailer_size(trailer);
+            assert(trailer_size >= parameters.offset);
```
We repeat similarly-shaped bit of code in superblock. Maybe we need a pair of helpers? 

```
self.superblock.trailer_data(trailer) -> []u8
self.superblock.trailer_data_sector_aligned(trailer) -> [][constants.sector_size]u8
```

(unrelated, but, yeah, I wonder if it makes sense to enforce, at compile time, that write_sectors accepts some number of sectrors, and not just []u8

sentientwaffle replied:
Hmm it seems like all of the uses are just slightly different :thinking: 
1. replica here just wants the bytes within the data; doesn't care about alignment
2. `superblock.write_trailer` wants to round up to the nearest sector -- but I think handling that explicitly there is a good thing, since it needs to be careful to zero out the sector padding
3. `superblock.read_trailer` wants to round up to the nearest sector

I think I'll leave it as-is.

matklad commented on src/vsr/superblock.zig:
```diff
@@ -1024,11 +1042,11 @@ pub fn SuperBlockType(comptime Storage: type) type {
+            const trailer = context.trailer.?;
 
-            context.write_trailer = null;
+            context.trailer = null;
```
Alternatively, instead of going `null -> .manifest -> null -> .free_set -> null -> .client_session -> null`, we could go `null -> .manifest -> .free_set -> .client_session -> null`, as that would narrow down the state transition somewhat.

sentientwaffle committed:
VSR/SuperBlock: Refactor write_trailers chain

- Make the `write_trailers`/`write_trailer_next` chain more clear.
- Rename `superblock.Area` to `SuperBlockZone`. (And rename its methods for clarity.)

sentientwaffle commented:
Oops I missed some of you comment somehow. Will update shortly.

sentientwaffle commented:
Ok, updated.

matklad commented on src/testing/storage.zig:
```diff
@@ -501,8 +501,8 @@ pub const Storage = struct {
     ) *const superblock.SuperBlockHeader {
-        const offset = vsr.Zone.superblock.offset(superblock.areas.header.offset(copy_));
-        const bytes = storage.memory[offset..][0..superblock.areas.header.size_max];
+        const offset = vsr.Zone.superblock.offset(superblock.SuperBlockZone.header.offset(copy_));
```
This shouldn't compile? We renamed `offset` I think?

sentientwaffle replied:
That is a `vsr.Zone`; I didn't rename `vsr.Zone.offset`.

sentientwaffle replied:
Oh wait, I didn't see the inner one
